### PR TITLE
Add `oadm migrate etcd-ttl` which encodes upstream TTL migration

### PR DESCRIPTION
**Pull request by smarterclayton**
This is in support of v2 -> v3

[test] @sdodson

### CODE REVIEW

**Comment by smarterclayton** [comment]
@ingvagabund taking the upstream and going to put it into openshift - we may need this in the future anyway

**Comment by smarterclayton** [comment]
Flake, wtf

Jun  9 20:26:32.845: INFO: At 2017-06-09 20:21:22 -0400 EDT - event
for nodeport-test-3z5g6: {kubelet nettest-node-2} MissingClusterDNS:
kubelet does not have ClusterDNS IP configured and cannot create Pod
using "ClusterFirst" policy. Falling back to DNSDefault policy.

**Comment by smarterclayton** [comment]
[test]

File: pkg/cmd/admin/migrate/etcd/ttl.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	lease, err := client.Lease.Grant(ctx, int64(o.leaseDuration/time.Second))
+	if err != nil {
+		return fmt.Errorf("unable to create lease: %v", err)
+	}
+	fmt.Fprintf(o.Out, "info: Lease with TTL %d created\n", lease.TTL)
+
+	fmt.Fprintf(o.Out, "info: Attaching lease to %d entries\n", len(objectsResp.Kvs))
+	errors := 0
+	for _, kv := range objectsResp.Kvs {
+		_, err := client.KV.Put(ctx, string(kv.Key), string(kv.Value), clientv3.WithLease(lease.ID))
```
Comment (liggitt) [inline pkg/cmd/admin/migrate/etcd/ttl.go]: make conditional so we don't stomp a changed value, gather items we need to retry?

File: pkg/cmd/admin/migrate/etcd/ttl.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		o.ttlKeysPrefix += "/"
+	}
+	ctx := context.Background()
+
+	objectsResp, err := client.KV.Get(ctx, o.ttlKeysPrefix, clientv3.WithPrefix())
+	if err != nil {
+		return fmt.Errorf("unable to get objects to attach to the lease: %v", err)
+	}
+
+	lease, err := client.Lease.Grant(ctx, int64(o.leaseDuration/time.Second))
```
Comment (liggitt) [inline pkg/cmd/admin/migrate/etcd/ttl.go]: are there any downsides to associating a lot of items with a single lease?

File: pkg/cmd/admin/migrate/etcd/ttl.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+	options.Bind(cmd.Flags())
+
+	return cmd
+}
+
+func (o *MigrateTTLReferenceOptions) Bind(flag *pflag.FlagSet) {
+	flag.StringVar(&o.etcdAddress, "etcd-address", "", "Etcd address")
+	flag.StringVar(&o.ttlKeysPrefix, "ttl-keys-prefix", "", "Prefix for TTL keys")
+	flag.DurationVar(&o.leaseDuration, "lease-duration", time.Hour, "Lease duration (seconds granularity)")
```
Comment (liggitt) [inline pkg/cmd/admin/migrate/etcd/ttl.go]: I wouldn't set a default duration... I would require the caller to decide

File: pkg/cmd/admin/migrate/etcd/ttl.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func generateClientConfig(o *MigrateTTLReferenceOptions) (*clientv3.Config, error) {
+	if o.etcdAddress == "" {
+		return nil, fmt.Errorf("--etcd-address flag is required")
+	}
+	if o.ttlKeysPrefix == "" {
+		return nil, fmt.Errorf("--ttl-keys-prefix flag is required")
+	}
+
```
Comment (liggitt) [inline pkg/cmd/admin/migrate/etcd/ttl.go]: require duration to be > time.Second

File: pkg/cmd/admin/migrate/etcd/ttl.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	client, err := clientv3.New(*c)
+	if err != nil {
+		return fmt.Errorf("unable to create etcd client: %v", err)
+	}
+
+	// Make sure that ttlKeysPrefix is ended with "/" so that we only get children "directories".
+	if !strings.HasSuffix(o.ttlKeysPrefix, "/") {
+		o.ttlKeysPrefix += "/"
+	}
```
Comment (liggitt) [inline pkg/cmd/admin/migrate/etcd/ttl.go]: require duration to be > time.Second

**Comment by smarterclayton** [comment]
All comments addressed, rebasing on master and then will merge

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[severity:blocker]

**Comment by smarterclayton** [comment]
[test]

**Comment by sdodson** [comment]
@ingvagabund this should be in the next 3.6 build so you can switch to using this

**Comment by ingvagabund** [comment]
Will check repos and update. @smarterclayton @sdodson thanks

**Comment by sdodson** [comment]
Should be in 3.6.112

**Pull request merged**